File: sdk/python/alp_sdk/zk_proof.py
import hashlib
import hmac
import os
from datetime import datetime, timezone
from typing import Dict, Any, Optional

class ZKProof:
    def __init__(
        self,
        proof_id: str,
        statement: str,
        commitment: str,
        proof_hash: str,
        verified: bool = True,
        created_at: Optional[str] = None,
    ):
        self.id = proof_id
        self.statement = statement
        self.commitment = commitment
        self.proof_hash = proof_hash
        self.verified = verified
        self.created_at = created_at or datetime.now(timezone.utc).isoformat()

class ZKProofEngine:
    def generate_proof(self, proof_id: str, statement: str, secret_value: str) -> ZKProof:
        salt = os.urandom(16).hex()
        commitment = hmac.new(salt.encode("utf-8"), secret_value.encode("utf-8"), hashlib.sha256).hexdigest()
        
        raw = f"{statement}:{commitment}:{salt}"
        proof_hash_val = hashlib.sha256(raw.encode("utf-8")).hexdigest()
        combined_proof_hash = f"{salt}:{proof_hash_val}"

        return ZKProof(
            proof_id=proof_id,
            statement=statement,
            commitment=commitment,
            proof_hash=combined_proof_hash,
            verified=True,
        )

    def verify_proof(self, proof: ZKProof) -> bool:
        if not proof or not proof.proof_hash or not proof.commitment:
            return False

        parts = proof.proof_hash.split(":")
        if len(parts) != 2:
            return False

        salt, expected_hash = parts
        raw = f"{proof.statement}:{proof.commitment}:{salt}"
        computed_hash = hashlib.sha256(raw.encode("utf-8")).hexdigest()

        return computed_hash == expected_hash

File: parser/src/zk-proof.ts
import * as crypto from 'node:crypto';

export interface ZKProof {
  id: string;
  statement: string;
  commitment: string;
  proofHash: string;
  verified: boolean;
  createdAt: string;
}

export class ZKProofEngine {
  /**
   * Generate a zero-knowledge commitment and proof hash for a given statement
   * using a cryptographic HMAC-SHA256 salt & hash without revealing the secret value.
   */
  public generateProof(id: string, statement: string, secretValue: string): ZKProof {
    const salt = crypto.randomBytes(16).toString('hex');
    const commitment = crypto.createHmac('sha256', salt).update(secretValue).digest('hex');
    
    // Proof hash binds the statement and commitment together
    const proofHash = crypto
      .createHash('sha256')
      .update(`${statement}:${commitment}:${salt}`)
      .digest('hex');

    return {
      id,
      statement,
      commitment,
      proofHash: `${salt}:${proofHash}`,
      verified: true,
      createdAt: new Date().toISOString(),
    };
  }

  /**
   * Verify whether a ZKProof proofHash matches its commitment and statement.
   */
  public verifyProof(proof: ZKProof): boolean {
    if (!proof || !proof.proofHash || !proof.commitment) return false;

    const parts = proof.proofHash.split(':');
    if (parts.length !== 2) return false;

    const [salt, expectedHash] = parts;
    const computedHash = crypto
      .createHash('sha256')
      .update(`${proof.statement}:${proof.commitment}:${salt}`)
      .digest('hex');

    return computedHash === expectedHash;
  }
}
